readthedocs/api/v3/mixins.py

```diff
@@ -1,12 +1,52 @@
 from django.shortcuts import get_object_or_404
 from rest_framework import status
 from rest_framework.response import Response
+from simple_history.utils import update_change_reason
 
 from readthedocs.builds.models import Version
 from readthedocs.organizations.models import Organization
 from readthedocs.projects.models import Project
 
 
+class UpdateChangeReasonMixin:
+
+    """
+    Set the change_reason on the model changed through this API view.
+
+    The view should inherit one of:
+
+    - CreateModelMixin
+    - UpdateModelMixin
+    - DestroyModelMixin
+
+    Unlike the original methods,
+    these return the instance that was created/updated,
+    so they are easy to override without having to save the object twice.
+    """
+
+    change_reason = None
+
+    def get_change_reason(self):
+        if self.change_reason:
+            return self.change_reason
+        klass = self.__class__.__name__
+        return f'origin=api-v3 class={klass}'
+
+    def perform_create(self, serializer):
+        obj = serializer.save()
+        update_change_reason(obj, self.get_change_reason())
+        return obj
+
+    def perform_update(self, serializer):
+        obj = serializer.save()
+        update_change_reason(obj, self.get_change_reason())
```

Comment:
I'm blocking the PR on exception handling on each `update_change_reason`. From my previous experience with EthicalAds, and your testing, I'm pretty confident this will cause production issues and raise 500's for users. We should roll this out with code that protects against this, unless there's some strong argument not to. A warning in Sentry is much better than a 500 for a user and a hotfix deploy.

Reply:
changed, what I don't like is that now tests won't catch these, as we are just logging now.

Reply:
We can write tests to catch this logic, can't we? Making the code raise more exceptions so that it's easier to test is definitely a trade off against making the code less prone to breaking in production. I don't feel comfortable making that trade, as we can certainly find ways to catch these log messages in tests can't we?